contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,245 @@
+#[cfg(feature = "redis_pool")]
+pub extern crate r2d2_redis;
+
+/// A struct containing database configuration options.
+///
+/// This retrieves the key-value pairs specified in the `Rocket.toml` under
+/// the databases key for use in further configuring the desired connection.
+///
+/// Example configuration:
+/// ```toml
+/// [[global.databases]]
+/// name = "my_database"
+/// url = "postgres://postgres@localhost/my_database
+/// pool_size = 4
+/// ```
+pub struct DatabaseConfig<'a> {
+    /// The connection URL specified in the Rocket configuration
+    pub url: &'a str,
+    /// The size of the pool to be initialized. Defaults to the number of Rocket workers.
+    pub pool_size: u32,
+    /// Any extra options that are included in the configuration.
+    pub extras: &'a BTreeMap<String, Value>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the R2D2 connection pool methods.
+#[derive(Debug)]
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+/// The error type for fetching the DatabaseConfig
+#[derive(Debug)]
+pub enum DatabaseConfigError {
+    MissingTable,
+    MalformedConfiguration,
+    MissingConnectionUrl,
+    InvalidConnectionUrl,
+    InvalidPoolSize(i64),
+}
+
+impl<'a> Display for DatabaseConfigError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            DatabaseConfigError::MissingTable => {
+                write!(f, "Database-specific configuration was not found for this database")
+            },
+            DatabaseConfigError::MalformedConfiguration => {
+                write!(f, "The configuration for this database is malformed")
+            }
+            DatabaseConfigError::MissingConnectionUrl => {
+                write!(f, "The connection URL is missing for this database")
+            },
+            DatabaseConfigError::InvalidConnectionUrl => {
+                write!(f, "The specified connection URL is invalid")
+            },
+            DatabaseConfigError::InvalidPoolSize(invalid_size) => {
+                write!(f, "'{}' is not a valid value for `pool_size`", invalid_size)
+            },
+        }
+    }
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+///
+/// Example usage:
+/// ```rust,ignore
+/// let config = database_config("my_database", rocket.config());
+/// ```
+pub fn database_config<'a>(
+    name: &str,
+    from: &'a config::Config
+) -> Result<DatabaseConfig<'a>, DatabaseConfigError> {
+    // Find the first `databases` config that's a table with a key of 'name'
+    // equal to `name`.
+    let connection_config = from.get_table("databases")
+        .map_err(|_| DatabaseConfigError::MissingTable)?
+        .get(name)
+        .iter()
+        .filter_map(|config| config.as_table())
+        .next()
```

Comment:
`iter/filter_map/next` can be replaced with `.and_then(|config| config.as_table())`. But to differentiate between missing entry and wrong data type, neither will work.